### PR TITLE
8303150: DCmd framework unnecessarily creates a DCmd instance on registration

### PR DESCRIPTION
**Pull request by dholmes-ora**
When DCmd factories are registered, the factory is passed the number of arguments taken by the DCmd - using a template method `get_num_arguments`. For DCmds that don't extend DCmdWithParser there has to be a static `num_arguments()` method in that class. For DCmds that do extend DCmdWithParser the logic instantiates an instance of the DCmd, extracts its parser and calls its `num_arguments` method which dynamically counts the number of defined arguments.

Creating an instance of each DCmd and dynamically counting arguments is inefficient and unnecessary, the number of arguments is statically known and easily expressed (in fact many of the JFR DCmds already statically define this). So we add the static `num_arguments()` method to each class that needs it and return the statically counted number of arguments. To ensure the static number and actual number don't get out-of-sync, we keep the original dynamic logic for use in debug builds to assert that the static and dynamic counts are the same. The assert will trigger during a debug build if something does get out of sync, for example if a new DCmd (extending DCmdWithParser) were added but didn't define the static `num_arguments()` method.

A number of DCmd classes were unnecessarily defining their own dynamic version of `num_arguments` and these are now removed.

In the template method I use `ENABLE_IF(std::is_convertible<T, DCmd>::value)` to check we only call this on DCmd classes. This may be unnecessary but it seemed consistent with the other template methods. Note that `std::is_base_of` only works for immediate super types.

Testing: tiers 1-4

Performance: in theory we should see some improvement in startup; in practice it is barely noticeable.

Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303150](https://bugs.openjdk.org/browse/JDK-8303150): DCmd framework unnecessarily creates a DCmd instance on registration

### Reviewers
 * [Frederic Parain](https://openjdk.org/census#fparain) (@fparain - Committer) ⚠️ Review applies to [5ef9724d](https://git.openjdk.org/jdk/pull/12994/files/5ef9724d96706b3ec7dea17e0e0f7d31c2822fae)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer) ⚠️ Review applies to [cfe345d8](https://git.openjdk.org/jdk/pull/12994/files/cfe345d8d2664cc98b90f5558d694d37fe086a70)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12994/head:pull/12994` \
`$ git checkout pull/12994`

Update a local copy of the PR: \
`$ git checkout pull/12994` \
`$ git pull https://git.openjdk.org/jdk pull/12994/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12994`

View PR using the GUI difftool: \
`$ git pr show -t 12994`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12994.diff">https://git.openjdk.org/jdk/pull/12994.diff</a>

</details>

### CODE REVIEW

**Review by fparain — APPROVED** [review]
The concern during the initial implementation was that the value returned by num_arguments() and the real number of arguments could become out of sync, but your solution to check that they are consistent only on debug builds addresses this concern.
Thank you for fixing this!

**Comment by dholmes-ora** [comment]
Thanks for the review @fparain !

File: src/hotspot/share/services/diagnosticFramework.hpp
Comment (kevinjwalls) [inline src/hotspot/share/services/diagnosticFramework.hpp:1]: Should we update this comment:
279   // - For subclasses of DCmdWithParser, it's calculated by DCmdParser::num_arguments().

Comment (dholmes-ora) [reply src/hotspot/share/services/diagnosticFramework.hpp:1]: Fixed - thanks.

File: src/hotspot/share/services/diagnosticFramework.hpp [hunk 1]
```diff
@@ -440,12 +438,12 @@ template <class DCmdClass> class DCmdFactoryImpl : public DCmdFactory {
...
 private:
   template <typename T, ENABLE_IF(!std::is_base_of<DCmdWithParser, T>::value)>
-  static int get_num_arguments() {
+  static int get_parsed_num_arguments() {
     return T::num_arguments();
   }
 
   template <typename T, ENABLE_IF(std::is_base_of<DCmdWithParser, T>::value)>
-  static int get_num_arguments() {
+  static int get_parsed_num_arguments() {
```
Comment (tstuefe) [inline src/hotspot/share/services/diagnosticFramework.hpp:446]: I don't understand why we need the dynamic allocation here. Should we not be able to call your new static versions of num_arguments like we do above? And if so, could we then not unify those two functions?

**Review by tstuefe — COMMENTED** [review]
Thanks a lot for removing that blemish, David! That has bugged me (and others) repeatedly.

Question inline, but looks good all in all.

**Comment by kevinjwalls** [comment]
Took me a bit of time, as I was thinking if we compare a hard-coded value with a parsed argument count, how does that cope with optional arguments?  Maybe "get_parsed_num_arguments()" suggested to me it was a live command being parsed, but DCmdParser's _arguments_list is a list of known possible arguments. Should have read your "number of arguments taken by the DCmd" at the top more carefully.

**Review by kevinjwalls — APPROVED** [review]

File: src/hotspot/share/services/diagnosticFramework.hpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/services/diagnosticFramework.hpp:446]: Thanks for looking at this @tstuefe. The whole point of the parsed version is to ensure the static version doesn't get forgotten or out-of-date.

Just realised the `get_parsed_num_arguments` should also be `ifdef ASSERT`.

**Comment by dholmes-ora** [comment]
Thanks for the review @kevinjwalls .

**Review by tstuefe — APPROVED** [review]
@dholmes-ora thanks for the explanation and the added ASSERT. LGTM

**Comment by dholmes-ora** [comment]
Thanks for the review @tstuefe . I will wait for some final build checks and then integrate.

**Pull request closed without merging**